src/main/java/fishman/utils/Parser.java

```diff
@@ -0,0 +1,131 @@
+package fishman.utils;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+import fishman.command.AddCommand;
+import fishman.command.Command;
+import fishman.command.DeleteCommand;
+import fishman.command.ExitCommand;
+import fishman.command.FindCommand;
+import fishman.command.ListCommand;
+import fishman.command.MarkCommand;
+import fishman.exception.FishmanException;
+import fishman.task.Deadline;
+import fishman.task.Event;
+import fishman.task.TaskList;
+import fishman.task.ToDo;
+
+
+/**
+ * The Parser class is used to interpret user input and create the appropriate
+ * Command objects.
+ */
+public class Parser {
+    private static final DateTimeFormatter INPUT_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
+
+    /**
+     * Parses the user input and returns the appropriate Command object.
+     * This method interprets the user input, splitting the input to determine the command
+     * and respective index and creates the appropriate
+     * Command object. It can identify the following:
+     * - "bye": Creates a ExitCommand which terminates the program.
+     * - "list": Creates a ListCommand to display all tasks in the task list.
+     * - "mark": Creates a MarkCommand flagged as mark to mark the task as done.
+     * - "unmark": Creates a MarkCommand flagged as unmark to mark the task as not done.
+     * - "to-do": Creates an AddCommand which creates a new To-Do task with the information provided by the user.
+     * - "deadline": Creates an AddCommand which creates a new Deadline task with the information provided by the user.
+     * - "event": Creates an AddCommand which creates a new Event task with the information provided by the user.
+     * - "delete": Creates a DeleteCommand which deletes the task.
+     * - "find": Creates a FindCommand which matches the tasks based on the keyword provided by the user.
+     * - Any other input: Creates a Command which will terminate the program.
+     *
+     * @param userInput The string input by the user.
+     * @return A Command object with respect to the user's input.
+     * @throws FishmanException.InvalidCommandException if input does not match any command.
+     * @throws FishmanException.MissingArgumentException if command is missing arguments.
+     * @throws FishmanException.NumberFormatException if a numeric argument is not a valid number.
+     * @throws FishmanException.EmptyListException if trying to mark/unmark a task in an empty task list.
+     * @throws FishmanException.IndexOutOfBoundsException if trying to choose a nonexistent item from the list.
+     * @throws FishmanException for unexpected errors during parsing.
+     */
+    public static Command parse(String userInput, TaskList tasks) throws FishmanException {
```

Comment:
There is some duplication in the code when checking the task list size and parsing indices for `mark`, `unmark`, and `delete`. Consider extracting this logic into a helper method to reduce redundancy.